src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * SupportedDependencyByCitus returns whether citus has support to distribute the object
+ * addressed.
+ */
+static bool
+SupportedDependencyByCitus(const ObjectAddress *address)
+{
+	/*
+	 * looking at the type of a object to see if we know how to create the object on the
+	 * workers.
+	 */
+	switch (getObjectClass(address))
+	{
+		case OCLASS_SCHEMA:
+		{
+			return true;
+		}
+
+		default:
+		{
+			/* unsupported type */
+			return false;
+		}
+	}
+
+	/*
+	 * all types should have returned above, compilers complaining about a non return path
+	 * indicate a bug in the above switch. Missing returns or breaking code flow should be
+	 * addressed in the switch statement, preferably not here.
+	 */
+}
+
+
+/*
+ * IsObjectAddressOwnedByExtension returns whether or not the object is owned by an
+ * extension. It is assumed that an object having a dependency on an extension is created
+ * by that extension and therefore owned by that extension.
+ */
+static bool
+IsObjectAddressOwnedByExtension(const ObjectAddress *target)
+{
+	Relation depRel = NULL;
+	ScanKeyData key[2] = { 0 };
+	SysScanDesc depScan = NULL;
+	HeapTuple depTup = NULL;
+	bool result = false;
+
+	depRel = heap_open(DependRelationId, AccessShareLock);
+
+	/* scan pg_depend for classid = $1 AND objid = $2 using pg_depend_depender_index */
+	ScanKeyInit(&key[0], Anum_pg_depend_classid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->classId));
+	ScanKeyInit(&key[1], Anum_pg_depend_objid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->objectId));
+	depScan = systable_beginscan(depRel, DependDependerIndexId, true, NULL, 2, key);
+
+	while (HeapTupleIsValid(depTup = systable_getnext(depScan)))
+	{
+		Form_pg_depend pg_depend = (Form_pg_depend) GETSTRUCT(depTup);
+		if (pg_depend->deptype == DEPENDENCY_EXTENSION)
+		{
+			result = true;
+			break;
+		}
+	}
+
+	systable_endscan(depScan);
+	relation_close(depRel, AccessShareLock);
+
+	return result;
+}
+
+
+/*
+ * OrderObjectAddressListInDependencyOrder given a list of ObjectAddresses return a new
+ * list of the same ObjectAddresses ordered on dependency order with the objects without
+ * dependencies first.
+ *
+ * The algortihm traveses pg_depend in a depth first order starting at the first object in
+ * the provided list. By traversing depth first it will put the first dependency at the
+ * head of the list with dependencies depending on them later.
+ *
+ * If the object is already in the list it is skipped for traversal. This happens when an
+ * object was already added to the target list before it occurred in the input list.
+ */
+List *
+OrderObjectAddressListInDependencyOrder(List *objectAddressList)
+{
+	List *targetList = NIL;
+	ListCell *ojectAddressCell = NULL;
+	foreach(ojectAddressCell, objectAddressList)
+	{
+		ObjectAddress *objectAddress = (ObjectAddress *) lfirst(ojectAddressCell);
+		if (IsObjectAddressInList(objectAddress, targetList))
```

Comment:
I kind of wonder about the complexity of this algorithm. How long does it take when there are e.g. 10,000 
or 100,000 distributed objects? An example could be a user doing schema-per-tenant, which is generally a bad idea, but still quite a common model for multi-tenant apps on PostgreSQL.  (pre-Citus)

I think we could afford to solve it later, but if you have time consider using a hash table (HTAB) for the target list/set.

Reply:
Here is an example file with 100K objects: https://gist.github.com/onderkalaci/01b83abfc849995d406a292084391aa3#file-100k_objects-sql

As I noted somewhere else, creating types one-by-one is one of the main bottlenecks. 

The total time to get `GetDependenciesForObject` is `119 seconds` for `200501` objects during `create_distributed_table`. The total time for replicating the types (including the above): `265 seconds`.

So, adding a new node also takes roughly takes = ~`265 - 119` = 150 seconds

We could definitely improve that, but we can live with 2 minutes for processing time for so large dependencies. We could consider adding `NOTICE` while processing the dependencies.

Reply:
> I think we could afford to solve it later, but if you have time consider using a hash table (HTAB) for the target list/set.

As Marco noted, I'd be also OK to handle this in the future. But, we could consider opening an issue like `Improve processing times for distributed object creation` and list the items mentioned here -- and more if you have ideas. (a) This function can be refactored to use HTAB (b) We can batch the commands into a single command (or potentially multiple but each consists of N items) (c) Idea that needs to be proved: Slightly change `recurse_pg_depend ` to keep less open scans by fetching all the dependent objects once, and iterating over them (d) Execute command across workers in parallel, we now do it sequentially

Reply:
I was aware that this is a heavy call, not that it was this bad (119 seconds bad), wonder if that is primarily due to the invocation here or at other places, or something else completely.

A list is used here because the order is quite important for the creation order. I have considered augmenting the data structure with set (HTAB was the implementation I was thinking of) to use for existence checking. But only using a set would not work here as the order in the list is essential.

Reply:
One alternative we could test it out could be keep both the list and HTAB. Do the `IsObjectAddressInList ` via the hash. That said, we should test with some large examples before doing that such as the example files here: https://gist.github.com/onderkalaci/01b83abfc849995d406a292084391aa3

Reply:
Another option would be to use a bloomfilter before the scan. https://github.com/postgres/postgres/blob/8255c7a5eeba8f1a38b7a431c04909bde4f5e67d/src/include/lib/bloomfilter.h

Reply:
I like the idea, and maybe we can go there in the future.
Yesterday we have implemented the HTAB solution and we went from 119 seconds to sth like 2-3 seconds. If in the future the memory consumption would be a problem for the set we can revisit the bloom filter approach.